repose-aggregator/core/repose-core/src/main/scala/org/openrepose/powerfilter/ReposeRoutingServlet.scala

```diff
@@ -0,0 +1,362 @@
+    defaultDestination.set(Option(interrogator.getDefaultDestination(configurationObject).orElse(null)))
+    destinations.set(localCluster.get().map(_.getDestinations) match {
+      case Some(dests) => getDestinations(dests.getEndpoint.asScala.toList) ++ getDestinations(dests.getTarget.asScala.toList)
+      case None => Map.empty[String, Destination]
+    })
+    domains.set(Option(new Clusters(configurationObject)))
+
+    if (logger.underlying.isDebugEnabled) {
+      localCluster.get().foreach { cluster =>
+        val clusterId: String = cluster.getId
+        //Build a list of the nodes in this cluster, just so we know what we're doing
+        val clusterNodes = cluster.getNodes.getNode.asScala.map(node => s"${node.getId}-${node.getHostname}")
+        logger.debug("{}:{} - Nodes for this router: {}", clusterId, nodeId, clusterNodes)
+        val destStrings = destinations.get()
+          .filter { case (_, dest) => dest.isInstanceOf[DestinationEndpoint] }
+          .map { case (_, endpoint) => s"${endpoint.getId}-${endpoint.asInstanceOf[DestinationEndpoint].getHostname}" }
+        logger.debug("{}:{} - Destinations for this router: {}", clusterId, nodeId, destStrings)
+      }
+    }
+    initialized = true
+
+    def getDestinations(destList: List[Destination]): Map[String, Destination] = {
+      destList.iterator.map(dest => dest.getId -> dest).toMap
+    }
+  }
+
+  override def isInitialized: Boolean = {
+    initialized
+  }
+
+  override def service(req: HttpServletRequest, resp: HttpServletResponse): Unit = {
+    if (!initialized) {
+      logger.error(NotInitializedMessage)
+      trySendError()
+    } else {
+      logger.trace("{} processing request...", this.getClass.getSimpleName)
+      Try(doRoute(new HttpServletRequestWrapper(req), resp)) recover {
+        case e =>
+          logger.error(FailedToRouteMessage, e)
+          trySendError()
+      }
+      logger.trace("{} returning response...", this.getClass.getSimpleName)
+    }
+
+    def trySendError(): Unit = {
+      if (!resp.isCommitted) {
+        resp.sendError(SC_INTERNAL_SERVER_ERROR, FailedToRouteMessage)
+      } else {
+        logger.error("Failed to Send Error due to response already being committed")
+      }
+    }
+  }
+
+  def doRoute(servletRequest: HttpServletRequestWrapper, servletResponse: HttpServletResponse): Unit = {
+    val attributeDestinations =
+      Option(servletRequest.getAttribute(CommonRequestAttributes.DESTINATIONS))
+        .getOrElse(List.empty[RouteDestination]).asInstanceOf[List[RouteDestination]]
+    val defaultDestinations = {
+      defaultDestination.get() match {
+        case Some(defDest) =>
+          List[RouteDestination] {
+            new RouteDestination(defDest.getId, servletRequest.getRequestURI, -1)
+          }
+        case _ => List.empty[RouteDestination]
+      }
+    }
+    (attributeDestinations ++ defaultDestinations)
+      .sortWith(_.compareTo(_) > 0) // Highest Quality first
```

Comment:
I think you can use `.sorted.reverse` here instead of this. Might be simpler?

Reply:
That ends up causing the list to be iterated over twice.
I wanted a `.lastOption`, but that doesn't exist.

Reply:
That thought did cross my mind, but I figured that this list would be small enough that we wouldn't care.

... but if we do, how about:

```
.sorted(Ordering[RouteDestination].reverse)
```

Reply:
It shouldn't be large, but if it were this would be an obscure thing to find.
Since it is easily avoidable I would rather sort it just once.
The `Ordering.reverse` is in essence doing the same thing, but I agree it is more readable.
Updated.